### PR TITLE
feat: retry only on specific exit code

### PR DESCRIPTION
**Pull request by nick-fields**
I just pulled the changes from the forked PR #57 into a branch so that Actions and tests would run.

### CODE REVIEW

File: .github/workflows/ci_cd.yml [hunk 1]
```diff
@@ -79,6 +78,42 @@ jobs:
...
+      - uses: nick-invision/assert-action@v1
+        with:
+          expected: failure
+          actual: ${{ steps.retry_on_exit_code_expected.outcome }}
+      - uses: nick-invision/assert-action@v1
+        with:
+          expected: 3
+          actual: ${{ steps.retry_on_exit_code_expected.outputs.total_attempts }}
+
+      - name: retry_on_exit_code (with unexpected error code)
```
Comment (nick-fields) [inline .github/workflows/ci_cd.yml:99]: @andersfischernielsen @erwinw

Can you take a look at [this](https://github.com/nick-fields/retry/runs/6077868023?check_suite_focus=true#step:19:32) test failure?  I made a few tweaks to your tests, first being that a duplicate `id` was used which caused the entire workflow to fail, then I added assertions to your two tests.

The assertion is failing for your `retry_on_exit_code (with unexpected error code)` test.  I assert that if an unexpected exit code is thrown (exit code 1), and the `retry_on_exit_code` is set (exit code 2), then the rerun should only occur once not 3x as is happening currently.  I _think_ you need to also handle `max_attempts`  when `retry_on_exit_code` is set and skip additional attempts if an unexpected exit code is returned.

Comment (erwinw) [reply .github/workflows/ci_cd.yml:99]: Hey @nick-fields , well danged, I've ~poked~ gently asked @andersfischernielsen to have a look at this!

**Comment by andersfischernielsen** [comment]
@nick-fields Sorry about the duped IDs 😬 

I’m pretty sure (take this with a grain of salt due to tired brain) that moving the check for the specific exit code to a separate `else if` would take care of the issue we’re seeing. AFAICT the existing return code checks interfere with just giving up and not retrying on our provided retry exit code 🤔 

Basically what I'd change the implementation to is:
image
As far as I understand the pre-existing code, this would allow us to blindly throw if a specific exit code is given _and_ we don't see that given exit code after the first failure, in turn leading to no retries. LMK if I'm misunderstanding the behaviour here 🙏🏼 

Great that moving the PR allows Actions to run, but I'm unable to contribute to this PR due to missing access rights. How would you like me to contribute a fix to the failing test?

**Comment by nick-fields** [comment]
> @nick-fields Sorry about the duped IDs 😬
> 
> I’m pretty sure (take this with a grain of salt due to tired brain) that moving the check for the specific exit code to a separate `else if` would take care of the issue we’re seeing. AFAICT the existing return code checks interfere with just giving up and not retrying on our provided retry exit code 🤔
> 
> In other words, what I'd change the implementation to is:
> 
> image
> As far as I understand the pre-existing code, this would allow us to blindly throw if a specific exit code is given _and_ we don't see that given exit code after the first failure, in turn leading to no retries. LMK if I'm misunderstanding the behaviour here 🙏🏼
> 
> Great that moving the PR allows Actions to run, but I'm unable to contribute to this PR due to missing access rights. How would you like me to contribute a fix to the failing test?

It was easier to just pull your proposed fix in and test it.  Looks like it worked! Merging now, and thanks for the contribution.

**Pull request merged**